File: cmd/watching.go
package cmd

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/mturley/agent-handler/config"
	"github.com/mturley/agent-handler/db"
	watcherPkg "github.com/mturley/agent-handler/watcher"
	"github.com/spf13/cobra"
)

var watchingCmd = &cobra.Command{
	Use:   "watching",
	Short: "Show watched resources, watcher status, and recent errors",
	RunE:  runWatching,
}

func init() {
	watchingCmd.GroupID = "human"
	rootCmd.AddCommand(watchingCmd)
	watchingCmd.Flags().String("session-id", "", "session ID (auto-detected if omitted)")
	watchingCmd.Flags().Bool("global", false, "show all watched resources across all sessions")
}

func runWatching(cmd *cobra.Command, args []string) error {
	d, err := openReadOnlyDB()
	if err != nil {
		return fmt.Errorf("failed to open database: %w", err)
	}
	defer d.Close()

	global, _ := cmd.Flags().GetBool("global")
	if global {
		return runWatchingGlobal(d)
	}

	sessionID, err := resolveSessionID(cmd)
	if err != nil {
		return fmt.Errorf("could not determine session: %w", err)
	}

	// Subscriptions
	subs, _ := d.ListSubscriptions(sessionID, false)

	// Watcher status
	cfg, _ := config.Read(config.DefaultPath())

	type watcherStatus struct {
		Name             string `json:"name"`
		Configured       bool   `json:"configured"`
		Installed        bool   `json:"installed"`
		Running          bool   `json:"running"`
		LastRun          string `json:"last_run,omitempty"`
		NextRun          string `json:"next_run,omitempty"`
		IntervalSeconds  int    `json:"interval_seconds,omitempty"`
		HasError         bool   `json:"has_error"`
		LastErrorMessage string `json:"last_error_message,omitempty"`
	}

	var watchers []watcherStatus
	for _, name := range []string{"github", "jira"} {
		ws := watcherStatus{Name: name}
		if cfg != nil {
			ws.Configured = cfg.IsServiceConfigured(name)
		}
		ws.Installed = watcherPkg.IsInstalled(name)
		ws.Running = watcherPkg.IsRunning(name)
		if lastRun := watcherPkg.LastRunTime(name); lastRun != nil {
			ws.LastRun = lastRun.Format(time.RFC3339)
			interval := watcherPkg.InstalledInterval(name)
			if interval > 0 {
				ws.IntervalSeconds = interval
				nextRun := lastRun.Add(time.Duration(interval) * time.Second)
				ws.NextRun = nextRun.Format(time.RFC3339)
			}
		}
		ws.HasError = d.HasWatcherError(name)
		if ws.HasError {
			if dbStatus, err := d.GetWatcherStatus(name); err == nil && dbStatus != nil {
				ws.LastErrorMessage = dbStatus.LastErrorMessage
			}
		}
		watchers = append(watchers, ws)
	}

	if jsonOutput {
		output := map[string]interface{}{
			"subscriptions": subs,
			"watchers":      watchers,
		}
		data, _ := json.MarshalIndent(output, "", "  ")
		fmt.Println(string(data))
		return nil
	}

	// Text output
	if len(subs) == 0 {
		fmt.Println("No resources are currently being watched in this session.")
	} else {
		fmt.Printf("Watched resources (%d):\n", len(subs))
		for _, sub := range subs {
			url := ""
			if sub.ResourceURL != nil {
				url = fmt.Sprintf("  %s", *sub.ResourceURL)
			}
			fmt.Printf("  %s:%s%s\n", sub.ResourceType, sub.ResourceID, url)
		}
	}

	fmt.Println()
	fmt.Println("Watchers:")
	for _, ws := range watchers {
		if !ws.Configured {
			fmt.Printf("  %s: not configured\n", ws.Name)
			continue
		}
		if !ws.Installed {
			fmt.Printf("  %s: configured but not installed\n", ws.Name)
			continue
		}
		lastRun := "never"
		nextRun := ""
		if ws.LastRun != "" {
			if t, err := time.Parse(time.RFC3339, ws.LastRun); err == nil {
				lastRun = formatDuration(time.Since(t)) + " ago"
			}
		}
		if ws.NextRun != "" {
			if t, err := time.Parse(time.RFC3339, ws.NextRun); err == nil {
				if t.After(time.Now()) {
					nextRun = fmt.Sprintf(", next: %s", formatDuration(time.Until(t)))
				} else {
					nextRun = ", next: any moment"
				}
			}
		}
		state := "running"
		if !ws.Running {
			state = "stopped"
		}
		if ws.HasError {
			fmt.Printf("  %s: %s, last run %s%s — ERROR\n", ws.Name, state, lastRun, nextRun)
			if ws.LastErrorMessage != "" {
				fmt.Printf("    %s\n", ws.LastErrorMessage)
			}
		} else {
			fmt.Printf("  %s: %s, last run %s%s\n", ws.Name, state, lastRun, nextRun)
		}
	}

	return nil
}

func runWatchingGlobal(d *db.DB) error {
	rows, err := d.Conn().Query(`
		SELECT DISTINCT sub.resource_type, sub.resource_id, COALESCE(sub.resource_url, ''),
			GROUP_CONCAT(DISTINCT COALESCE(s.session_name, s.branch)) as sessions
		FROM subscriptions sub
		JOIN sessions s ON s.session_id = sub.session_id
		WHERE sub.deleted_at IS NULL AND s.status = 'active'
		GROUP BY sub.resource_type, sub.resource_id
		ORDER BY sub.resource_type, sub.resource_id
	`)
	if err != nil {
		return fmt.Errorf("failed to query subscriptions: %w", err)
	}
	defer rows.Close()

	type globalSub struct {
		ResourceType string `json:"resource_type"`
		ResourceID   string `json:"resource_id"`
		ResourceURL  string `json:"resource_url,omitempty"`
		Sessions     string `json:"sessions"`
	}

	var results []globalSub
	for rows.Next() {
		var gs globalSub
		rows.Scan(&gs.ResourceType, &gs.ResourceID, &gs.ResourceURL, &gs.Sessions)
		results = append(results, gs)
	}

	// Watcher status (same as per-session view)
	cfg, _ := config.Read(config.DefaultPath())
	type wsStatus struct {
		Name             string `json:"name"`
		Configured       bool   `json:"configured"`
		Installed        bool   `json:"installed"`
		Running          bool   `json:"running"`
		LastRun          string `json:"last_run,omitempty"`
		NextRun          string `json:"next_run,omitempty"`
		HasError         bool   `json:"has_error"`
		LastErrorMessage string `json:"last_error_message,omitempty"`
	}
	var watchers []wsStatus
	for _, name := range []string{"github", "jira"} {
		ws := wsStatus{Name: name}
		if cfg != nil {
			ws.Configured = cfg.IsServiceConfigured(name)
		}
		ws.Installed = watcherPkg.IsInstalled(name)
		ws.Running = watcherPkg.IsRunning(name)
		if lastRun := watcherPkg.LastRunTime(name); lastRun != nil {
			ws.LastRun = lastRun.Format(time.RFC3339)
			interval := watcherPkg.InstalledInterval(name)
			if interval > 0 {
				nextRun := lastRun.Add(time.Duration(interval) * time.Second)
				ws.NextRun = nextRun.Format(time.RFC3339)
			}
		}
		ws.HasError = d.HasWatcherError(name)
		if ws.HasError {
			if dbStatus, err := d.GetWatcherStatus(name); err == nil && dbStatus != nil {
				ws.LastErrorMessage = dbStatus.LastErrorMessage
			}
		}
		watchers = append(watchers, ws)
	}

	if jsonOutput {
		output := map[string]interface{}{
			"subscriptions": results,
			"watchers":      watchers,
		}
		data, _ := json.MarshalIndent(output, "", "  ")
		fmt.Println(string(data))
		return nil
	}

	if len(results) == 0 {
		fmt.Println("No active subscriptions across any session")
	} else {
		fmt.Printf("Watched resources across all sessions (%d):\n\n", len(results))
		for _, gs := range results {
			fmt.Printf("  %s:%s\n", gs.ResourceType, gs.ResourceID)
			if gs.ResourceURL != "" {
				fmt.Printf("    URL: %s\n", gs.ResourceURL)
			}
			if gs.Sessions != "" {
				fmt.Printf("    Sessions: %s\n", gs.Sessions)
			}
			fmt.Println()
		}
	}

	fmt.Println("Watchers:")
	for _, ws := range watchers {
		if !ws.Configured {
			fmt.Printf("  %s: not configured\n", ws.Name)
			continue
		}
		if !ws.Installed {
			fmt.Printf("  %s: configured but not installed\n", ws.Name)
			continue
		}
		lastRun := "never"
		nextRun := ""
		if ws.LastRun != "" {
			if t, err := time.Parse(time.RFC3339, ws.LastRun); err == nil {
				lastRun = formatDuration(time.Since(t)) + " ago"
			}
		}
		if ws.NextRun != "" {
			if t, err := time.Parse(time.RFC3339, ws.NextRun); err == nil {
				if t.After(time.Now()) {
					nextRun = fmt.Sprintf(", next: %s", formatDuration(time.Until(t)))
				} else {
					nextRun = ", next: any moment"
				}
			}
		}
		state := "running"
		if !ws.Running {
			state = "stopped"
		}
		if ws.HasError {
			fmt.Printf("  %s: %s, last run %s%s — ERROR\n", ws.Name, state, lastRun, nextRun)
			if ws.LastErrorMessage != "" {
				fmt.Printf("    %s\n", ws.LastErrorMessage)
			}
		} else {
			fmt.Printf("  %s: %s, last run %s%s\n", ws.Name, state, lastRun, nextRun)
		}
	}

	return nil
}

File: src/Handler/Cmd/Watching.hs
-- | Port of cmd/watching.go: show watched resources, watcher status, and
-- recent errors.
module Handler.Cmd.Watching (watchingCommand) where

import Control.Monad (forM, forM_, unless)
import Data.Aeson (Value, object, toJSON, (.=))
import Data.Maybe (catMaybes, fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time.Clock (UTCTime, addUTCTime)
import Data.Time.Format (defaultTimeLocale, formatTime)
import qualified Database.SQLite.Simple as SQL
import Options.Applicative

import Handler.Cli.Common
import Handler.Cmd.Status (ansiBold, ansiDim, ansiGreen, ansiRed, ansiReset, ansiYellow, formatDuration, parseIso, sinceSeconds)
import Handler.Config (configDefaultPath, isServiceConfigured, readConfig)
import Handler.Db (Db(..), close)
import Handler.Db.Subscriptions (Subscription(..), listSubscriptions, subscriptionToJson)
import Handler.Db.WatcherStatus (WatcherStatus(..), getWatcherStatus, hasWatcherError)
import Handler.Watcher.Scheduler (installedInterval, isInstalled, isRunning, lastRunTime)

data WatchingOpts = WatchingOpts
  { wSessionId :: Maybe Text
  , wGlobal    :: Bool
  }

watchingCommand :: Mod CommandFields NamedCommand
watchingCommand = mkCommand "watching" "Show watched resources, watcher status, and recent errors" (runWatching <$> opts)
  where
    opts = WatchingOpts
      <$> sessionIdOption
      <*> switch (long "global" <> help "show all watched resources across all sessions")

-- | Watcher health summary row shared by both views.
data WatcherView = WatcherView
  { wvName       :: Text
  , wvConfigured :: Bool
  , wvInstalled  :: Bool
  , wvRunning    :: Bool
  , wvLastRun    :: Text
  , wvNextRun    :: Text
  , wvInterval   :: Int
  , wvHasError   :: Bool
  , wvErrMessage :: Text
  }

isoFormat :: UTCTime -> Text
isoFormat = T.pack . formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ"

buildWatcherViews :: Db -> IO [WatcherView]
buildWatcherViews db = do
  cfg <- configDefaultPath >>= readConfig
  forM ["github", "jira"] $ \name -> do
    let configured = isServiceConfigured cfg name
    installed <- isInstalled name
    running <- isRunning name
    mLastRun <- lastRunTime name
    (lastRun, nextRun, interval) <- case mLastRun of
      Nothing -> pure ("", "", 0)
      Just lr -> do
        interval <- installedInterval name
        let next = if interval > 0
              then isoFormat (addUTCTime (fromIntegral interval) lr)
              else ""
        pure (isoFormat lr, next, interval)
    hasErr <- hasWatcherError db name
    errMsg <-
      if hasErr
        then maybe "" (.lastErrorMessage) <$> getWatcherStatus db name
        else pure ""
    pure WatcherView
      { wvName = name, wvConfigured = configured, wvInstalled = installed
      , wvRunning = running, wvLastRun = lastRun, wvNextRun = nextRun
      , wvInterval = interval, wvHasError = hasErr, wvErrMessage = errMsg
      }

watcherViewJson :: Bool -> WatcherView -> Value
watcherViewJson includeInterval wv = object $
  [ "name" .= wv.wvName
  , "configured" .= wv.wvConfigured
  , "installed" .= wv.wvInstalled
  , "running" .= wv.wvRunning
  , "has_error" .= wv.wvHasError
  ]
  ++ [ "last_run" .= wv.wvLastRun | wv.wvLastRun /= "" ]
  ++ [ "next_run" .= wv.wvNextRun | wv.wvNextRun /= "" ]
  ++ [ "interval_seconds" .= wv.wvInterval | includeInterval && wv.wvInterval > 0 ]
  ++ [ "last_error_message" .= wv.wvErrMessage | wv.wvErrMessage /= "" ]

-- | \"5m ago\" / \"never\" and \", next: 2m\" / \", next: any moment\" strings.
lastNextStrings :: WatcherView -> IO (Text, Text)
lastNextStrings wv = do
  lastRun <- case parseIso wv.wvLastRun of
    Just t | wv.wvLastRun /= "" -> do
      since <- sinceSeconds t
      pure (formatDuration since <> " ago")
    _ -> pure "never"
  nextRun <- case parseIso wv.wvNextRun of
    Just t | wv.wvNextRun /= "" -> do
      since <- sinceSeconds t
      pure $ if since < 0
        then ", next: " <> formatDuration (negate since)
        else ", next: any moment"
    _ -> pure ""
  pure (lastRun, nextRun)

runWatching :: WatchingOpts -> Ctx -> IO ()
runWatching o ctx = do
  db <- openReadOnlyDb
  if o.wGlobal
    then runWatchingGlobal db ctx
    else do
      sessionId <- resolveSessionIdOpt o.wSessionId
      subs <- listSubscriptions db sessionId False
      watchers <- buildWatcherViews db

      if ctx.jsonOutput
        then printJson $ object
          [ "subscriptions" .= map subscriptionToJson subs
          , "watchers" .= map (watcherViewJson True) watchers
          ]
        else do
          if null subs
            then putTextLn "No resources are currently being watched in this session."
            else do
              printfT ["Watched resources (", T.pack (show (length subs)), "):"]
              forM_ subs $ \sub -> do
                let url = maybe "" ("  " <>) sub.resourceUrl
                printfT ["  ", sub.resourceType, ":", sub.resourceId, url]

          putTextLn ""
          putTextLn "Watchers:"
          forM_ watchers $ \wv ->
            if | not wv.wvConfigured -> printfT ["  ", wv.wvName, ": not configured"]
               | not wv.wvInstalled -> printfT ["  ", wv.wvName, ": configured but not installed"]
               | otherwise -> do
                   (lastRun, nextRun) <- lastNextStrings wv
                   let state = if wv.wvRunning then "running" else "stopped"
                   if wv.wvHasError
                     then do
                       printfT ["  ", wv.wvName, ": ", state, ", last run ", lastRun, nextRun, " — ERROR"]
                       unless (wv.wvErrMessage == "") $
                         printfT ["    ", wv.wvErrMessage]
                     else printfT ["  ", wv.wvName, ": ", state, ", last run ", lastRun, nextRun]
  close db

runWatchingGlobal :: Db -> Ctx -> IO ()
runWatchingGlobal db ctx = do
  resRows <- SQL.query_ db.conn
    "SELECT DISTINCT sub.resource_type, sub.resource_id, COALESCE(sub.resource_url, '')\
    \ FROM subscriptions sub\
    \ JOIN sessions s ON s.session_id = sub.session_id\
    \ WHERE sub.deleted_at IS NULL AND s.status = 'active'\
    \ ORDER BY sub.resource_type, sub.resource_id"
    :: IO [(Text, Text, Text)]

  results <- forM resRows $ \(rType, rId, rUrl) -> do
    sessRows <- SQL.query db.conn
      "SELECT s.session_id, COALESCE(s.session_name, ''), s.branch, s.last_active\
      \ FROM subscriptions sub\
      \ JOIN sessions s ON s.session_id = sub.session_id\
      \ WHERE sub.resource_type = ? AND sub.resource_id = ?\
      \   AND sub.deleted_at IS NULL AND s.status = 'active'\
      \ ORDER BY s.last_active DESC"
      (rType, rId)
      :: IO [(Text, Text, Text, Text)]
    pure (rType, rId, rUrl, sessRows)

  watchers <- buildWatcherViews db

  if ctx.jsonOutput
    then printJson $ object
      [ "subscriptions" .= toJSON
          [ object $
              [ "resource_type" .= rType
              , "resource_id" .= rId
              , "sessions" .= toJSON
                  [ object $
                      [ "session_id" .= sid, "branch" .= branch, "last_active" .= lastActive ]
                      ++ [ "session_name" .= name | name /= "" ]
                  | (sid, name, branch, lastActive) <- sessions ]
              ]
              ++ [ "resource_url" .= rUrl | rUrl /= "" ]
          | (rType, rId, rUrl, sessions) <- results ]
      , "watchers" .= map (watcherViewJson False) watchers
      ]
    else do
      if null results
        then putTextLn "No active subscriptions across any session"
        else do
          printfT ["Watched resources across all sessions (", T.pack (show (length results)), "):\n"]
          forM_ results $ \(rType, rId, rUrl, sessions) -> do
            printfT ["  ", ansiBold, rType, ":", rId, ansiReset]
            unless (rUrl == "") $
              printfT ["  ", ansiDim, rUrl, ansiReset]
            forM_ sessions $ \(sid, name0, branch, lastActive0) -> do
              let name = if name0 == "" then branch else name0
              lastActive <- case parseIso lastActive0 of
                Just t -> do
                  since <- sinceSeconds t
                  pure (formatDuration since <> " ago")
                Nothing -> pure lastActive0
              printfT ["  ", ansiDim, "└ ", name, " (", T.take 12 sid, ") — ", lastActive, ansiReset]
            putTextLn ""

      printfT [ansiDim, "─── Watchers ───", ansiReset]
      forM_ watchers $ \wv ->
        if | not wv.wvConfigured ->
               printfT ["  ", wv.wvName, ansiRed, "✗", ansiReset, " ", ansiDim, "not configured", ansiReset]
           | not wv.wvInstalled ->
               printfT ["  ", wv.wvName, " ", ansiYellow, "✗", " ", ansiReset, " ", ansiDim, "(not installed)", ansiReset]
           | otherwise -> do
               (lastRun, nextRun) <- lastNextStrings wv
               if | wv.wvHasError -> do
                      printfT ["  ", ansiRed, "✗", ansiReset, " ", wv.wvName, " ", ansiDim,
                               "(last run: ", lastRun, nextRun, ")", ansiReset]
                      unless (wv.wvErrMessage == "") $
                        printfT ["  ", ansiDim, "  ", wv.wvErrMessage, ansiReset]
                  | not wv.wvRunning ->
                      printfT ["  ", ansiYellow, "⏸", ansiReset, " ", wv.wvName, " ", ansiDim,
                               "(stopped, last run: ", lastRun, ")", ansiReset]
                  | otherwise ->
                      printfT ["  ", ansiGreen, "✓", ansiReset, " ", wv.wvName, " ", ansiDim,
                               "(last run: ", lastRun, nextRun, ")", ansiReset]
